### PR TITLE
tools: Add watch mode to webpack-make

### PR DESCRIPTION
**Pull request by martinpitt**
Watch mode makes fast iterations much faster. Compared to plain `make`,
watch mode and disabling eslint reduces e. g. tuned from 19s to 3s, or
systemd from 48s to 17s on the first rebuild, and 5s (!) on the second.

### CODE REVIEW

**Comment by martinpitt** [comment]
Whoops, I still had another PR locally applied; fixed

**Comment by martinpitt** [comment]
Resolved conflict and dropped erroneous whitespace change.

**Comment by martinpitt** [comment]
Issues from IRC:

 * Enable polling if `stat -fc %T .` == `nfs`
 * Don't exit on errors

File: HACKING.md [hunk 1]
```diff
@@ -155,8 +155,14 @@ Use the same user and password that you used to log into your Linux desktop.
...
 http://localhost:9090
 
-After every change to your sources, run `make` to update all the webpacks, and
-reload cockpit in your browser.
+After every change to your sources the webpacks need to be rebuilt: You can
+just run run `make` to update everything that has changed; for iterating
+faster, you can run webpack in "watch" mode on the particular page that you are
+working on, which reduces the build time to less than a third. E. g.
+
+    $ ESLINT=0 tools/webpack-make -d dist/systemd/Makefile.deps -w
```
Comment (marusak) [inline HACKING.md:163]: hmm, so I cannot just watch it all? I need to be specific?

**Comment by mvollmer** [comment]
>  * Enable polling if `stat -fc %T .` == `nfs`

This is not important anymore, since I will not be running `webpack-make -w` inside the VM.  Not using a VM is just fine and much faster.

**Comment by mvollmer** [comment]
>  * Don't exit on errors

I feel this is important.  If I have to baby-sit the watch process, it's just as distracting as waiting a minute for the results, maybe even more so bc I get so irritated... :-)

File: HACKING.md (hunk 1 above)
Comment (martinpitt) [reply HACKING.md:163]: Yes, both a direct `webpack`, `webpack --watch`, or `tools/webpack-make` without `-d` is utterly broken right now. I have not yet looked into why, or whether this is even supposed to work. For now we only support building them individually. I figure it's also what you want for fast iteration -- you don't want to rebuild everything when you touch a file in pkg/lib/, for example.

**Comment by marusak** [comment]
> Don't exit on errors

Absolutely. The 20s advantage disappears immediately when I don't realize that it failed and will wait for too long until it propagates :)
I liked the old '--watch' where one would just start it and hack on whatever pages and it would just rebuild it. And when things got broken, it would show big red error, but still would keep watching.

**Comment by mvollmer** [comment]
> And when things got broken, it would show big red error, but still would keep watching.

Except it would often break down and not actually recover from the error, at which point you need to restart it, IIRC. Hopefully it's better now.

**Comment by mvollmer** [comment]
I have pushed two commits, and with those I am happy I guess.

Fundamentally, I want incremental repacking but I don't care much about the file watching. I want to have a webpack process in the background for the page that I am working on, and I want to send it a signal of some sort every now and then when I want it to do its thing.

But if watching is the price to pay to get incremental builds, that's fine.

File: HACKING.md [hunk 2]
```diff
@@ -155,8 +155,14 @@ Use the same user and password that you used to log into your Linux desktop.
 
 http://localhost:9090
 
-After every change to your sources, run `make` to update all the webpacks, and
-reload cockpit in your browser.
+After every change to your sources the webpacks need to be rebuilt: You can
+just run run `make` to update everything that has changed; for iterating
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply HACKING.md]: Fixed, thanks for spotting!

**Comment by martinpitt** [comment]
I reworked this again to incorporate @mvollmer's fixups (I reworked webpack-watch a bit, though), added a "build started" message, simplified the output format, and addressed the typo from above. PTAL?

**Pull request merged**